Check if worker is initialized

nkaretnikov opened the pull request:
Fixes #605.

## Description

conda-store can run via docker and in standalone mode:
- in standalone mode, the `--standalone` CLI parameter must be used to start the worker sub-process
- when running via docker, this CLI parameter is not used because a dedicated worker process is started by docker-compose.

However, there might be cases where users forget to pass the `--standalone` parameter when running in standalone mode. This is a problem because the server cannot execute any tasks without the worker, so it must be running.

This PR adds a check that verifies whether the worker is running and prints an error message if not.

The worker is configured via a number of parameters, like `celery_broker_url`, which can either point to a database connection or use Redis:

```py
    @default("celery_broker_url")
    def _default_celery_broker_url(self):
        if self.redis_url is not None:
            return self.redis_url
        return f"sqla+{self.database_url}"
```

Similarly, checks like `conda_store.celery_app.control.inspect().ping()` won't work outside of Redis and RabbitMQ.

So we use a database entry to check whether the worker is initialized. The database is used because it will work regardless of parameters used to configure the worker.

Example output:
```
[CondaStoreServer] WARNING | Waiting for worker... Use --standalone if running outside of docker
[CondaStoreServer] WARNING | Waiting for worker... Use --standalone if running outside of docker
[CondaStoreServer] WARNING | Waiting for worker... Use --standalone if running outside of docker
[CondaStoreServer] Worker initialized
```

The warning message is printed in red, the info message is printed in green.

The simplest way to test this:
- run the server in one terminal, wait for warnings to be printed
- run the worker in another terminal, wait for the info message to be printed.

The commands to run from `conda-store/conda-store-server`:
```
terminal1: python -m conda_store_server.server
terminal2: python -m conda_store_server.worker
```

To test in standalone mode (also starts the worker):
```
python -m conda_store_server.server --standalone
```

To test via docker:
```
docker-compose build --build-arg RELEASE_VERSION=2023.10.1 && docker-compose up
```

## Pull request checklist

<!-- Quick checklist to ensure high-quality Pull Request. -->

- [x] Did you test this change locally?
- [ ] Did you update the documentation (if required)?
- [ ] Did you add/update relevant tests for this change (if required)?

## Additional information

<!-- Do you have any other information about this pull request? This may include screenshots, references, and/or implementation notes. -->

nkaretnikov commented on conda-store-server/conda_store_server/server/app.py:
```diff
@@ -200,6 +203,12 @@ def initialize(self, *args, **kwargs):
+        # self.conda_store.celery_app
+        # import time
+        # time.sleep(5)
+        # assert self.conda_store.celery_app.control.inspect().ping() is not None
```
Debug, remove

nkaretnikov commented on conda-store-server/conda_store_server/server/app.py:
```diff
@@ -340,16 +349,53 @@ async def favicon():
 
+            # Needs to delete this entry to check if the worker is available
+            db.query(orm.Worker).delete()
+            db.commit()
```
This needs to happen on the server side before the worker is started to avoid a race condition. Moreover, it needs to happen on the server side because it will work in case the server is terminated and there are existing entries in the DB from previous runs.

nkaretnikov replied:
I need to add this as a comment in the code

nkaretnikov commented on conda-store-server/conda_store_server/server/app.py:
```diff
@@ -340,16 +349,53 @@ async def favicon():
+                    "Waiting for worker... "
+                    "Use --standalone if running outside of docker"
+                    f"{reset}"
+                )
```
Added a signal handler because uvicorn ignores all signals in the start event. You wouldn't be able to terminate this loop otherwise.

nkaretnikov replied:
Also add a comment for this

nkaretnikov commented on conda-store-server/conda_store_server/worker/tasks.py:
```diff
@@ -46,6 +47,19 @@ def worker(self):
 
 
+# Signals to the server that the worker is running, see startup_event in
+# CondaStoreServer.init_fastapi_app
```
This comment needs to be updated. I've renamed the functions

nkaretnikov commented on conda-store-server/conda_store_server/server/app.py:
```diff
@@ -343,13 +346,73 @@ async def favicon():
+                    "Waiting for worker... "
+                    "Use --standalone if running outside of docker"
+                    f"{reset}"
+                )
```
Try using `self.log` here if it works

nkaretnikov commented on conda-store-server/conda_store_server/server/app.py:
```diff
@@ -28,6 +28,9 @@
 from traitlets.config import Application, catch_config_error
 
+logger = logging.getLogger("app")
+logger.setLevel(logging.INFO)
```
This needs to be updated to use `basicConfig` because this uses the implicit handler.

nkaretnikov commented:
Tested this on Linux and Windows. Example output from Windows (server and worker are running in separate terminals):

<img width="1350" alt="Screen Shot 2023-12-18 at 23 25 02" src="https://github.com/conda-incubator/conda-store/assets/10469885/90a973c9-3111-4273-b4fd-419fcfc52a04">

dcmcand commented on conda-store-server/conda_store_server/server/app.py:
```diff
@@ -343,13 +343,73 @@ async def favicon():
+
+        def check_worker():
+            # Uses colors to make the output more visible
+            green = "\x1b[32m"
```
I feel like these should be constants. There are likely going to be other places that we want to use those colors and not redefining them would be nice.

dcmcand commented on conda-store-server/conda_store_server/server/app.py:
```diff
@@ -343,13 +343,73 @@ async def favicon():
         return app
 
     def start(self):
+        import time
```
Why are you importing these here instead of with all the other imports?

dcmcand commented on conda-store-server/conda_store_server/server/app.py:
```diff
@@ -343,13 +343,73 @@ async def favicon():
+                        self.log.info(f"{green}" "Worker initialized" f"{reset}")
+                        break
+
+                time.sleep(5)
```
It would be nice to have this be passed in as a parameter with a default value so that for testing we could make the time shorter.

dcmcand commented on conda-store-server/conda_store_server/orm.py:
```diff
@@ -41,6 +42,22 @@
+
+    __table_args__ = (
+        # Ensures no duplicates can be added with this combination of fields.
+        UniqueConstraint("initialized", name="_uc_worker"),
```
This unique constraint means that if you ever initialize more than one worker, then you get sql errors. You can try this by starting two workers and you will see:
'''
[2024-01-05 09:16:14,152: ERROR/ForkPoolWorker-3] Task task_initialize_worker[62967c9a-aca7-4f00-b056-15ec3a606507] raised unexpected: IntegrityError('(sqlite3.IntegrityError) UNIQUE constraint failed: worker.initialized')
Traceback (most recent call last):
  File "/home/chuck/.micromamba/envs/conda-store-server/lib/python3.10/site-packages/sqlalchemy/engine/base.py", line 1900, in _execute_context
    self.dialect.do_execute(
  File "/home/chuck/.micromamba/envs/conda-store-server/lib/python3.10/site-packages/sqlalchemy/engine/default.py", line 736, in do_execute
    cursor.execute(statement, parameters)
sqlite3.IntegrityError: UNIQUE constraint failed: worker.initialized

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/chuck/.micromamba/envs/conda-store-server/lib/python3.10/site-packages/celery/app/trace.py", line 477, in trace_task
    R = retval = fun(*args, **kwargs)
  File "/home/chuck/.micromamba/envs/conda-store-server/lib/python3.10/site-packages/celery/app/trace.py", line 760, in __protected_call__
    return self.run(*args, **kwargs)
  File "/home/chuck/devel/conda-store/conda-store-server/conda_store_server/worker/tasks.py", line 61, in task_initialize_worker
    db.commit()
  File "/home/chuck/.micromamba/envs/conda-store-server/lib/python3.10/site-packages/sqlalchemy/orm/session.py", line 1451, in commit
    self._transaction.commit(_to_root=self.future)
  File "/home/chuck/.micromamba/envs/conda-store-server/lib/python3.10/site-packages/sqlalchemy/orm/session.py", line 829, in commit
    self._prepare_impl()
  File "/home/chuck/.micromamba/envs/conda-store-server/lib/python3.10/site-packages/sqlalchemy/orm/session.py", line 808, in _prepare_impl
    self.session.flush()
  File "/home/chuck/.micromamba/envs/conda-store-server/lib/python3.10/site-packages/sqlalchemy/orm/session.py", line 3444, in flush
    self._flush(objects)
  File "/home/chuck/.micromamba/envs/conda-store-server/lib/python3.10/site-packages/sqlalchemy/orm/session.py", line 3583, in _flush
    with util.safe_reraise():
  File "/home/chuck/.micromamba/envs/conda-store-server/lib/python3.10/site-packages/sqlalchemy/util/langhelpers.py", line 70, in __exit__
    compat.raise_(
  File "/home/chuck/.micromamba/envs/conda-store-server/lib/python3.10/site-packages/sqlalchemy/util/compat.py", line 211, in raise_
    raise exception
  File "/home/chuck/.micromamba/envs/conda-store-server/lib/python3.10/site-packages/sqlalchemy/orm/session.py", line 3544, in _flush
    flush_context.execute()
  File "/home/chuck/.micromamba/envs/conda-store-server/lib/python3.10/site-packages/sqlalchemy/orm/unitofwork.py", line 456, in execute
    rec.execute(self)
  File "/home/chuck/.micromamba/envs/conda-store-server/lib/python3.10/site-packages/sqlalchemy/orm/unitofwork.py", line 630, in execute
    util.preloaded.orm_persistence.save_obj(
  File "/home/chuck/.micromamba/envs/conda-store-server/lib/python3.10/site-packages/sqlalchemy/orm/persistence.py", line 245, in save_obj
    _emit_insert_statements(
  File "/home/chuck/.micromamba/envs/conda-store-server/lib/python3.10/site-packages/sqlalchemy/orm/persistence.py", line 1238, in _emit_insert_statements
    result = connection._execute_20(
  File "/home/chuck/.micromamba/envs/conda-store-server/lib/python3.10/site-packages/sqlalchemy/engine/base.py", line 1705, in _execute_20
    return meth(self, args_10style, kwargs_10style, execution_options)
  File "/home/chuck/.micromamba/envs/conda-store-server/lib/python3.10/site-packages/sqlalchemy/sql/elements.py", line 334, in _execute_on_connection
    return connection._execute_clauseelement(
  File "/home/chuck/.micromamba/envs/conda-store-server/lib/python3.10/site-packages/sqlalchemy/engine/base.py", line 1572, in _execute_clauseelement
    ret = self._execute_context(
  File "/home/chuck/.micromamba/envs/conda-store-server/lib/python3.10/site-packages/sqlalchemy/engine/base.py", line 1943, in _execute_context
    self._handle_dbapi_exception(
  File "/home/chuck/.micromamba/envs/conda-store-server/lib/python3.10/site-packages/sqlalchemy/engine/base.py", line 2124, in _handle_dbapi_exception
    util.raise_(
  File "/home/chuck/.micromamba/envs/conda-store-server/lib/python3.10/site-packages/sqlalchemy/util/compat.py", line 211, in raise_
    raise exception
  File "/home/chuck/.micromamba/envs/conda-store-server/lib/python3.10/site-packages/sqlalchemy/engine/base.py", line 1900, in _execute_context
    self.dialect.do_execute(
  File "/home/chuck/.micromamba/envs/conda-store-server/lib/python3.10/site-packages/sqlalchemy/engine/default.py", line 736, in do_execute
    cursor.execute(statement, parameters)
sqlalchemy.exc.IntegrityError: (sqlite3.IntegrityError) UNIQUE constraint failed: worker.initialized
[SQL: INSERT INTO worker (initialized) VALUES (?)]
[parameters: (1,)]
(Background on this error at: https://sqlalche.me/e/14/gkpj)
'''

I suggest a better approach would be to remove the unique constraint and simply query the worker table for initialized = true. If there is at least one row that is returned then would can proceed.

nkaretnikov commented on conda-store-server/conda_store_server/server/app.py:
```diff
@@ -350,6 +387,30 @@ def start(self):
             self.conda_store.ensure_namespace(db)
             self.conda_store.ensure_conda_channels(db)
 
+            # This ensures the database has no Worker table entires when the
```
Typo: entires, will fix

nkaretnikov commented:
@dcmcand Please take a look!

Summary of changes:
- The constants and method are accessible globally, but they are marked as private to avoid polluting the public namespace.
- The time delay value is now a parameter.
- Removed the DB constraint, too.
- Fixed a typo.

dcmcand reviewed (approved):
:rocket: Thanks @nkaretnikov
